### ISSUE TITLE
Tasking Manager Load Testing

### ISSUE THREAD

**Issue by arunasank** [hotosm/tasking-manager#1354]
I load tested the Tasking Manager last week with locustio, and the findings have been pretty interesting.

## Locust?
Locust works by simulating requests to specific end points, and weighting requests by end points. The end points are defined in a file: [`locustfile.py`](https://github.com/hotosm/tasking-manager-locustio-test/blob/master/locustfile.py), and requests are sent to each of these end points.

## Load Test Simulation Settings
* Number of users: 500
* Hatch rate (Number of users added/s): 50/s

### Load Test 1
Initially, I tested on a single EC2 - a `c3.2xlarge`, which has 8 CPU cores, and 15GB RAM. Snippets from the results are as follows:

<img width="1224" alt="screen shot 2019-02-25 at 5 00 08 pm" src="https://user-images.githubusercontent.com/3166852/53337373-3d82c980-3927-11e9-9092-d29bb9c59772.png">
<img width="1216" alt="screen shot 2019-02-25 at 5 00 15 pm" src="https://user-images.githubusercontent.com/3166852/53337374-3d82c980-3927-11e9-8565-cdfd189de800.png">
<img width="1206" alt="screen shot 2019-02-25 at 5 00 20 pm" src="https://user-images.githubusercontent.com/3166852/53337375-3d82c980-3927-11e9-8718-f3a99229b44a.png">
<img width="1235" alt="screen shot 2019-02-25 at 5 00 33 pm" src="https://user-images.githubusercontent.com/3166852/53337376-3d82c980-3927-11e9-8d0b-544c889162f6.png">

### Load Test 2

Next, I increased the number of instances to 4, so requests are distributed equally among all instances. I also got rid of the requests to `/`. Snippets from the results are as follows:
<img width="674" alt="screen shot 2019-02-25 at 5 42 37 pm" src="https://user-images.githubusercontent.com/3166852/53337410-58553e00-3927-11e9-87bf-e85671eb982c.png">
Continued....
<img width="677" alt="screen shot 2019-02-25 at 6 02 55 pm" src="https://user-images.githubusercontent.com/3166852/53337528-b2ee9a00-3927-11e9-9952-99deb5dd3dbb.png">

## Next actions
* The CPU utilization across instances is really low - it averages around 2 - 10%, even though I am running gunicorn with `(cores * 2)+1` workers. I also ran `htop` on these instances, and saw that all the CPU cores are not being used. This needs to be investigated - I am not sure if there is an option on gunicorn to spread workers across various processes. Testing async workers and optimising gunicorn may help with some of these issues.
* When the request count dips, and then increases sharply, it's accompanied by a massive spike in latency (as you can see in load test 1, where the latency increases to 720s at that yellow peak) - why does this happen? I think fixing 1, can help fix this issue as well.
* All the failures are a result of gateway timeouts - this is when the requests take longer than the ELB timeout period.
* Running tests on eu-west-1 vs us-east-1.

## Locust 
* I need to figure out getting locust stats across a time range, and having a file with these stats
* A single locust instance, that is localhost is sufficient to run queries for our scale, but this is also likely influenced by network and bandwidth locally. Going to set this up on an ec2, once I figure out a way to get the output stats.

cc/ @hotosm/tech

**Comment by arunasank** [comment]
The results of the new load tests after implementing some asynchronous behaviour in gunicorn look so good! 😊 We are using the same instance type as what is currently on production, for the cloudformation load test as well. 

# Production metrics

### Request distribution
```
Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
GET /                                                          14728     5(0.03%)     320       6   59680  |      89    4.90
GET /contribute                                                14706     4(0.03%)     337       6   59994  |      90    5.00
GET /contribute?difficulty=ALL                                 14666    11(0.07%)     363       5   59776  |      89    3.60
GET /project/5207                                              14478     8(0.06%)     358       5   60005  |      90    4.00
GET /project/5236                                              14656    11(0.07%)     369       6   59901  |      90    6.10
--------------------------------------------------------------------------------------------------------------------------------------------
Total                                                          73234    39(0.05%)                                      23.60
```

### Percentage of the requests completed within given times
```
Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
GET /                                                           14728     89    130    170    200    350    500   2200   7500  60000
GET /contribute                                                 14706     90    130    170    200    350    520   2300   7800  60000
GET /contribute?difficulty=ALL                                  14666     89    130    170    210    360    530   2600   7700  60000
GET /project/5207                                               14478     90    130    180    210    360    550   3500   8000  60000
GET /project/5236                                               14656     90    130    170    210    360    530   3100   8000  60000
--------------------------------------------------------------------------------------------------------------------------------------------
Total                                                           73234     89    130    170    210    350    530   2800   7800  60000
```

# Cloudformation metrics

### Request distribution

```
 Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                          43617     1(0.00%)      55       4    2868  |      33   53.50
 GET /contribute                                                43163     1(0.00%)      56       4    2114  |      33   49.70
 GET /contribute?difficulty=ALL                                 43186     2(0.00%)      56       5    2842  |      33   51.70
 GET /project/5207                                              43326     1(0.00%)      56       4    2104  |      33   52.20
 GET /project/5236                                              43409     2(0.00%)      55       4    2858  |      33   52.20
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                         216701     7(0.00%)                                     259.30
```

### Percentage of the requests completed within given times
```
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                           43617     33     37     39     41     47     79    250   1000   2900
 GET /contribute                                                 43163     33     37     39     41     47     88    260    980   2100
 GET /contribute?difficulty=ALL                                  43186     33     37     39     41     47     92    260   1000   2800
 GET /project/5207                                               43326     33     37     39     41     47     83    260   1100   2100
 GET /project/5236                                               43409     33     37     39     41     47     73    260    990   2900
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                          216701     33     37     39     41     47     83    260   1000   2900
```

**Comment by arunasank** [comment]
Based on https://github.com/hotosm/tasking-manager/issues/1445, I ran out some fresh load tests on the CloudFormation tasking manager and the Production tasking manager. Here are the results for a load test were 50 users are using the tasking manager ans 20 new users are accessing the tasking manager per second.

In my first load test, I ran out the tests in this way:

## CloudFormation stack: 1 instance with 2 CPU cores

```
Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                          45562     0(0.00%)     419       4   26892  |     410   11.90
 GET /api/v1/project/5824                                       45907     0(0.00%)     593     137   26977  |     570   12.80
 GET /api/v1/project/5824/mapped-tasks-by-user                  45691     0(0.00%)     428      11    2473  |     420   13.00
 GET /contribute                                                45217     1(0.00%)     420       4   26931  |     410    9.70
 GET /contribute?difficulty=ALL                                 44995     0(0.00%)     420       4   26899  |     410   10.90
 GET /learn                                                     45453     2(0.00%)     420       3   52491  |     410   12.10
 GET /project/5207                                              45432     0(0.00%)     420       4   26653  |     410   12.00
 GET /project/5236                                              45277     0(0.00%)     421       4   26789  |     410   13.50
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                         363534     3(0.00%)                                      95.90

Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                           45562    410    510    570    610    730    840    970   1000  27000
 GET /api/v1/project/5824                                        45907    570    680    740    790    910   1000   1200   1300  27000
 GET /api/v1/project/5824/mapped-tasks-by-user                   45691    420    520    580    610    740    850    980   1000   2500
 GET /contribute                                                 45217    410    510    570    600    730    840    960   1000  27000
 GET /contribute?difficulty=ALL                                  44995    410    510    570    600    730    840    970   1000  27000
 GET /learn                                                      45453    410    510    570    610    730    840    970   1000  52000
 GET /project/5207                                               45432    410    510    570    610    730    840    970   1000  27000
 GET /project/5236                                               45277    410    510    570    610    730    840    970   1000  27000
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                          363534    420    540    590    640    760    870   1000   1100  52000
```

## Production stack: 1 instance with 2 CPU cores

```
Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                          24716     0(0.00%)     653       4   27825  |      48    0.00
 GET /api/v1/project/5824                                       24793     0(0.00%)    1102     165   27824  |     420    0.00
 GET /api/v1/project/5824/mapped-tasks-by-user                  24768 24768(50.00%)     639       8   27411  |      55    0.00
 GET /contribute                                                25478     0(0.00%)     651       5   27826  |      49    0.00
 GET /contribute?difficulty=ALL                                 24812     0(0.00%)     654       4   27460  |      47    0.00
 GET /learn                                                     24670     0(0.00%)     669       5   27503  |      49    0.00
 GET /project/5207                                              24909     0(0.00%)     621       4   27729  |      50    0.00
 GET /project/5236                                              24761     0(0.00%)     663       5   27825  |      49    0.00
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                         198907 24768(12.45%)                                       0.00

Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                           24716     48    120    180    230    390   1100  14000  14000  28000
 GET /api/v1/project/5824                                        24793    420    510    600    680   1000   1600  14000  15000  28000
 GET /api/v1/project/5824/mapped-tasks-by-user                   24768     55    130    190    230    410   1000  14000  14000  27000
 GET /contribute                                                 25478     49    120    180    230    400   1100  14000  14000  28000
 GET /contribute?difficulty=ALL                                  24812     47    120    180    230    400   1100  14000  14000  27000
 GET /learn                                                      24670     49    120    180    230    400   1100  14000  14000  28000
 GET /project/5207                                               24909     50    120    180    230    390   1000  14000  14000  28000
 GET /project/5236                                               24761     49    120    180    230    400   1100  14000  14000  28000
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                          198907     76    180    290    340    590   1200  14000  14000  28000
```

The results, while good were not so promising. Then, I realised that the capacity on the production stack was higher at 3 instances, so I repeated my load test with three instances on the CloudFormation stack as well.

**Comment by arunasank** [comment]
## CloudFormation Stack: 3 instances with 2 CPU cores each

```
Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                          42636    87(0.20%)     373       3  109897  |      44    0.00
 GET /api/v1/project/5824                                       42771    74(0.17%)     775       3  132493  |     340    0.00
 GET /api/v1/project/5824/mapped-tasks-by-user                  42370   104(0.24%)     359       7   90078  |      60    0.00
 GET /contribute                                                42754    64(0.15%)     329       3  108274  |      46    0.00
 GET /contribute?difficulty=ALL                                 42339    82(0.19%)     360       4  109900  |      48    0.00
 GET /learn                                                     42308    93(0.22%)     328       3   91397  |      47    0.00
 GET /project/5207                                              42495    83(0.19%)     323       3  108473  |      46    0.00
 GET /project/5236                                              42540   108(0.25%)     350       4  109529  |      45    0.00
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                         340213   695(0.20%)                                       0.00

Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                           42636     44    140    190    230    400    710   1500   1800 110000
 GET /api/v1/project/5824                                        42771    340    470    620    740   1300   1600   2000   2400 132000
 GET /api/v1/project/5824/mapped-tasks-by-user                   42370     60    150    200    250    420    770   1500   1700  90000
 GET /contribute                                                 42754     46    140    190    230    390    680   1500   1700 108000
 GET /contribute?difficulty=ALL                                  42339     48    140    190    230    390    700   1500   1800 110000
 GET /learn                                                      42308     47    140    190    230    400    710   1500   1700  91000
 GET /project/5207                                               42495     46    140    180    230    390    650   1400   1700 108000
 GET /project/5236                                               42540     45    140    190    230    400    700   1500   1700 110000
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                          340213     91    180    250    310    520   1000   1600   1900 132000

Error report
 # occurrences      Error
--------------------------------------------------------------------------------------------------------------------------------------------
 17                 GET /learn: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 12                 GET /: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 88                 GET /project/5236: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/project/5236',)"
 12                 GET /contribute?difficulty=ALL: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 66                 GET /api/v1/project/5824: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/api/v1/project/5824',)"
 84                 GET /api/v1/project/5824/mapped-tasks-by-user: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/api/v1/project/5824/mapped-tasks-by-user',)"
 71                 GET /project/5207: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/project/5207',)"
 12                 GET /project/5207: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 20                 GET /project/5236: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 20                 GET /api/v1/project/5824/mapped-tasks-by-user: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 70                 GET /contribute?difficulty=ALL: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/contribute?difficulty=ALL',)"
 16                 GET /contribute: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 8                  GET /api/v1/project/5824: "ConnectionError(ProtocolError('Connection aborted.', RemoteDisconnected('Remote end closed connection without response',)),)"
 48                 GET /contribute: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/contribute',)"
 76                 GET /learn: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/learn',)"
 75                 GET /: "HTTPError('502 Server Error: Bad Gateway for url: https://tasks-cfn-stage.hotosm.org/',)"
--------------------------------------------------------------------------------------------------------------------------------------------
```

## Analysis: At the Load Balancer

**Request count vs Latency**

<img width="1042" alt="Screen Shot 2019-04-17 at 6 54 33 PM" src="https://user-images.githubusercontent.com/3166852/56292704-4765c380-6145-11e9-8797-ee8fe688a785.png">

**Request count vs 5XX count**

All the 5XXs were 502s (The orange is not visible because it is overlaid on perfectly by the 502 count)

<img width="1044" alt="Screen Shot 2019-04-17 at 6 53 57 PM" src="https://user-images.githubusercontent.com/3166852/56292753-5c425700-6145-11e9-9def-6f86288ab5fb.png">

**Request Count vs 2XX count**
<img width="1048" alt="Screen Shot 2019-04-17 at 6 55 33 PM" src="https://user-images.githubusercontent.com/3166852/56292855-96abf400-6145-11e9-9df6-8646196a0183.png">

Note that the initial responses were 5XXs for a while, all 502s and then we see only 2XXs

## Analysis: Per Target Group

**Request Count per Target vs 5XX count**

<img width="1035" alt="Screen Shot 2019-04-17 at 6 58 55 PM" src="https://user-images.githubusercontent.com/3166852/56292952-c9ee8300-6145-11e9-8c94-aaaa7d7638d5.png">

Note that there are no 5XXs at the target level - which means that all the 502s were responses sent by the load balancer. This could be because we run long health checks on our targets, and we should reduce the time taken to finish the health checks.

**Request Count vs Healthy and Unhealthy Hosts**

<img width="1055" alt="Screen Shot 2019-04-17 at 7 00 10 PM" src="https://user-images.githubusercontent.com/3166852/56293066-028e5c80-6146-11e9-8d0c-9c013d456765.png">

What is also puzzling is that all the targets are called healthy before we start our load test.

**Comment by arunasank** [comment]
## Production Load Test: 3 instances with 2 CPU cores each

```
Name                                                          # reqs      # fails     Avg     Min     Max  |  Median   req/s
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                          15727     0(0.00%)    1082       5   13114  |     190    1.40
 GET /api/v1/project/5824                                       15571     0(0.00%)    1486     164   12996  |     560    1.40
 GET /api/v1/project/5824/mapped-tasks-by-user                  15541 15541(50.00%)    1076       8    8402  |     200    2.00
 GET /contribute                                                15675     0(0.00%)    1027       5   13116  |     190    1.10
 GET /contribute?difficulty=ALL                                 15651     0(0.00%)    1076       5   13036  |     190    0.50
 GET /learn                                                     15607     0(0.00%)    1071       4   11108  |     190    1.40
 GET /project/5207                                              15391     0(0.00%)    1082       4    7388  |     200    1.20
 GET /project/5236                                              15523     0(0.00%)    1067       4    8159  |     200    1.70
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                         124686 15541(12.46%)                                      10.70

Percentage of the requests completed within given times
 Name                                                           # reqs    50%    66%    75%    80%    90%    95%    98%    99%   100%
--------------------------------------------------------------------------------------------------------------------------------------------
 GET /                                                           15727    190    500    900   1100   5300   5600   5800   5900  13000
 GET /api/v1/project/5824                                        15571    560    850   1100   1500   5500   5700   5900   6000  13000
 GET /api/v1/project/5824/mapped-tasks-by-user                   15541    200    520    900   1100   5200   5500   5800   5900   8400
 GET /contribute                                                 15675    190    480    860   1000   5200   5500   5700   5900  13000
 GET /contribute?difficulty=ALL                                  15651    190    490    870   1100   5300   5600   5800   5900  13000
 GET /learn                                                      15607    190    490    870   1100   5300   5600   5800   5900  11000
 GET /project/5207                                               15391    200    530    900   1100   5300   5600   5800   5900   7400
 GET /project/5236                                               15523    200    530    900   1100   5200   5600   5800   5900   8200
--------------------------------------------------------------------------------------------------------------------------------------------
 Total                                                          124686    280    580    900   1100   5300   5600   5800   5900  13000

Error report
 # occurrences      Error
--------------------------------------------------------------------------------------------------------------------------------------------
 15541              GET /api/v1/project/5824/mapped-tasks-by-user: "HTTPError('404 Client Error: NOT FOUND for url: https://tasks.hotosm.org/api/v1/project/5824/mapped-tasks-by-user',)"
--------------------------------------------------------------------------------------------------------------------------------------------
```

## Analysis: At the Load Balancer

**Request count vs Latency**

<img width="1051" alt="Screen Shot 2019-04-17 at 8 34 13 PM" src="https://user-images.githubusercontent.com/3166852/56298683-35d5e900-6150-11e9-92e2-0476ce8ffbd0.png">


**Request count vs 5XX count**

<img width="1040" alt="Screen Shot 2019-04-17 at 8 31 21 PM" src="https://user-images.githubusercontent.com/3166852/56298597-0921d180-6150-11e9-99f4-e9204d3f0af3.png">

**Request Count vs 2XX count**
<img width="1055" alt="Screen Shot 2019-04-17 at 8 23 43 PM" src="https://user-images.githubusercontent.com/3166852/56298631-1dfe6500-6150-11e9-8f7e-db87c17afd6b.png">

## Analysis: Per Target Group

**Request Count vs Healthy and Unhealthy Hosts**

<img width="1045" alt="Screen Shot 2019-04-17 at 8 39 53 PM" src="https://user-images.githubusercontent.com/3166852/56299139-125f6e00-6151-11e9-9da5-1c2271b23d29.png">

This is the most puzzling bit for me since there's a mismatch between what's on the console vs CloudWatch. Trying to figure this out - any ideas @dakotabenjamin ?

**Three instances in service on the load balancer**
<img width="1056" alt="Screen Shot 2019-04-17 at 8 36 08 PM" src="https://user-images.githubusercontent.com/3166852/56299153-19867c00-6151-11e9-8fa9-7354be3a3e3f.png">

**Three instances in service on the ASG**
<img width="1031" alt="Screen Shot 2019-04-17 at 8 37 02 PM" src="https://user-images.githubusercontent.com/3166852/56299154-19867c00-6151-11e9-9dd1-d14e8375117e.png">

**Activity history shows that instances were created 3 hours before the load tests**
<img width="1036" alt="Screen Shot 2019-04-17 at 8 37 09 PM" src="https://user-images.githubusercontent.com/3166852/56299157-19867c00-6151-11e9-9068-80d97836d976.png">

## AutoScaling metrics

<img width="1055" alt="Screen Shot 2019-04-17 at 8 46 25 PM" src="https://user-images.githubusercontent.com/3166852/56299572-edb7c600-6151-11e9-9f46-adbd042885d0.png">

CPU utilisation was ~80% at peak, and no failed health checks, so again the mismatch in the two metrics is surprising.

**Comment by arunasank** [comment]
Overall, I feel really good about our new stack, and I think we are ready to move ahead based on the load tests. The latency looks sharp, and we are able to serve about 20-25 users per EC2 instance we are using. Which should be a good indicator of how to scale up our stacks during peak traffic.

@willemarcel the latency we saw during your load tests was because we were running fewer instances on the CloudFormation stack vs the production stack, which we have fixed now. Thanks for identifying that issue. 🙇‍♀️
